### PR TITLE
Cylindrical Cut Section or Cut Tube Implementation

### PR DESCRIPTION
**Pull request by ianna**
- DD definition and an interface to G4CutTubs shape
- Visualization of the shape in Fireworks
- Unit tests for the new shape

### CODE REVIEW

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
FYI, @davidlange6, @civanch, @friccita, @schneiml, @boudoul, @veszpv

This interface to G4CutTubs shape allows much simpler DD definition of Pixel support disks as 3 sections. It is a tube cut by two planes in Z. The cutting angles are defined by an outside normal (lx,ly,lz) at -z and an outside normal (tx,ty,tz) at +z : 
```
<CutTubs name="tube2" rMin="1*m" rMax="2*m" dz="1*m" startPhi="30*deg" deltaPhi="150*deg" lx="0" ly="-0.7" lz="-0.71" tx="0.7" ty="0" tz="0.71"/>
```

**Comment by civanch** [comment]
+1
worse trying, however, this shape is not used frequently, so if this shape will be used in pixel geometry, then CMS will be practically the first big consumer, at least, to my knowledge.

**Comment by davidlange6** [comment]
looks like its used here https://twiki.ph.rhul.ac.uk/twiki/bin/view/PP/JAI/BdSim - maybe worth talking with them about issues?
(G4CutTubs is not really so new, came in from g4.9.5 it seems)

On Aug 2, 2016, at 11:57 AM, Vladimir Ivantchenko <notifications@github.com<mailto:notifications@github.com>> wrote:


+1
worse trying, however, this shape is not used frequently, so if this shape will be used in pixel geometry, then CMS will be practically the first big consumer, at least, to my knowledge.

**Comment by ianna** [comment]
Here is an xml definition example: 
```
<SolidSection label="pixfwdInnerDiskSupportRings.xml">
...
<CutTubs name="PixelForwardInnerDiskInnerRingByYana"  startPhi="0." deltaPhi="360*deg"
   rMin="[InnerRingRMin]" rMax="[InnerRingCFRMax]" dz="[InnerRingHalfWidth]"
           lx="0" ly="-0.7" lz="-0.71"
           tx="0.7" ty="0" tz="0.71"/>
...
</SolidSection>

<LogicalPartSection label="pixfwdInnerDiskSupportRings.xml">
...
<LogicalPart name="PixelForwardInnerDiskInnerRingByYana" category="envelope">
  <rSolid name="PixelForwardInnerDiskInnerRingByYana"/>
  <rMaterial name="materials:Air"/>
 </LogicalPart>
...
</LogicalPartSection>

<PosPart copyNumber="2">
 <rParent name="pixfwdDisks:PixelForwardDisk1"/>
 <rChild name="pixfwdInnerDiskSupportRings:PixelForwardInnerDiskInnerRingByYana"/>
 <Translation x="0.*mm" y="0.*mm" z="[pixfwdInnerDiskSupportRings:InnerRingZ]"/>
</PosPart>
```
screen shot 2016-08-02 at 12 56 05

**Comment by ianna** [comment]
please test

**Comment by schneiml** [comment]
Wow. This seems useful. However, if my imagination does not fail me, this does not entirely solve the problem:

The support rings are (currently in out code and afaik also "in reality") defined piecewise linear in *phi*, while a tubs cut by planes will be piecewise linear in *x,y,z*. This means it would be off by some nonlinear amount.

I am not certain how the actual rings are defined, and an approximation is fine anyways, but we might still have to use a number of blocks with weird approximated parameters.

**Comment by ianna** [comment]
please test

**Comment by alja** [comment]
+1

**Comment by ianna** [comment]
@schneiml - Yes, one wouldn't expect to get away with one shape. However, looking at one half of the support disk, it seems to me it can be made of 3-4 cut tubes. One has to calculate the correct normals though ;-)

In any case it would be at least an order of magnitude less nodes for a DAG traversal with a plus of a more accurate description.

screen shot 2016-08-02 at 16 20 59
screen shot 2016-08-02 at 16 22 22
screen shot 2016-08-02 at 16 24 28

**Comment by schneiml** [comment]
@ianna yes, I am currently looking at the CAD file and came up with a 7-segment description (for one half ring, symmetry within one halfring does not really exist). That is quite a bit of work, but the Algorithm takes a similar number of parameters and we'd have to re-measure them anyways. Also this 7-segment solution would be fairly precise, especially at the "jump" where we have to be careful with overlapping the blades.

Also, it seems the one side of the ring is more axis-linear, the other side is more phi-linear, so it does not make a big difference. 

The next challenge is then computing the normals from the measurements...

**Comment by ianna** [comment]
@cmsbuild - please, check what's wrong with DAS

**Comment by ianna** [comment]
+1

**Comment by schneiml** [comment]
I have trouble to compile this code (`git cms-merge-topic` 'ed on `pre9`). On lxplus/slc6_amd64_gcc493 the SimG4Core package does not compile due to `-Werror=unused-variable`; I assume I can turn this off, but what is the right way to do this?

On a slc6_amd64_gcc530 install, it complains about a missing `G4EmParameters.hh`. I see this file in Geant4 `10.02.p02`, but not in the `10.00.p03-giojec3` that seems to be used.

Any hints?

**Comment by civanch** [comment]
@schneiml , you likely mix two versions of Geant4 when cms-merge-topic. To avoid this, start from one of recent IBs and not from pre9.

**Comment by ianna** [comment]
please test

**Comment by schneiml** [comment]
@civanch ah, that sounds plausible, thanks. Let me try.

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
@cmsbuild - it looks like the IB is broken.

**Comment by schneiml** [comment]
@ianna I see some unexpected behaviour here: the `deltaPhi` parameter is actually not a delta, but the end value. This is consistent with the root documentation, but does not really make sense given the name...

Idk whether the normal tubs behaves the same, but if not, this should probably be fixed.

**Comment by ianna** [comment]
@schneiml - thanks, I'll check.

**Comment by ianna** [comment]
Here is how G4 defines it:
```
G4CutTubs( const G4String& pName,
                    G4double pRMin,
                    G4double pRMax,
                    G4double pDz,
                    G4double pSPhi,
                    G4double pDPhi,
                    G4ThreeVector pLowNorm,
                    G4ThreeVector pHighNorm )

In the picture:
pRMin = 12, pRMax = 20, pDz = 30, pSPhi = 0, pDPhi = 1.5*pi, pLowNorm = (0,-0.7,-0.71), pHighNorm = (0.7,0,0.71)

giving its name pName and its parameters which are:

pRMin	Inner radius	
pRMax	Outer radius
pDz	        Half length in z	
pSPhi	Starting phi angle in radians
pDPhi	Angle of the segment in radians	
pLowNorm	Outside Normal at -z
pHighNorm	Outside Normal at +z	 	 
```

**Comment by schneiml** [comment]
Hmm... The thing that I know is that according to the Fireworks rendering, the `deltaPhi` in the XML is **not** a delta. It is the end value. If root and Geant4 are not consistent here, that is even more dangerous.

**Comment by ianna** [comment]
@davidlange6 - there is a problem with previously merged PR which is shown in DD unit tests:
```
===== Test "testmat" ====
cms::Exception caught in testmat
An exception of category 'FatalRootError' occurred while
   [0] Constructing service of type InitRootHandlers
   Additional Info:
      [a] Fatal Root Error: @SUB=TInterpreter::ReadRootmapFile
class  TotemRPDetId found in libDataFormatsTotemRPDetId.so  is already in libDataFormatsCTPPSDetId.so 

Done!!!

---> test testmat had ERRORS

```

**Comment by ianna** [comment]
@schneiml - you are right. Root definition is different:
```
- first phi limit (in degrees)
- second phi limit
 The segment will be be placed from the first angle (first phi limit)
 to the second angle (second phi limit)
```
I'll fix it in the DD to Root dump. The DD to G4 translation is correct.

File: Fireworks/Geometry/src/TGeoMgrFromDdd.cc [hunk 1]
```diff
@@ -263,6 +263,18 @@ TGeoMgrFromDdd::createShape(const std::string& iName,
...
 	    break;
+	 case ddcuttubs:
+	    //Order in params is  zhalf,rIn,rOut,startPhi,deltaPhi,lx,ly,lz,tx,ty,tz
+	    rSolid= new TGeoCtub(
+				 iName.c_str(),
+				 params[1]/cm,
+				 params[2]/cm,
+				 params[0]/cm,
+				 params[3]/deg,
+				 params[4]/deg,
```
Comment (ianna) [inline Fireworks/Geometry/src/TGeoMgrFromDdd.cc]: it should be:
```
params[3]/deg + params[4]/deg
```

**Comment by ianna** [comment]
please test

**Comment by schneiml** [comment]
@ianna looks good now.

one half ring

(also @veszpv )

**Comment by ianna** [comment]
+1

**Comment by ianna** [comment]
@davidlange6 - please, include this PR in the next IB. Both simulation and visualization have signed it. The test fails due to a DAS error.

**Pull request merged**